src/keywords/items.rs

```diff
@@ -22,103 +25,101 @@ impl ItemsArrayValidator {
+        instance_value
+            .iter()
+            .zip(self.items.iter())
+            .all(move |(item, validators)| {
+                validators
+                    .iter()
+                    .all(move |validator| helpers::is_valid(validator.deref(), schema, item))
+            })
     }
 
-    fn name(&self) -> String {
-        format!("items: [{}]", format_vec_of_validators(&self.items))
+    fn validate_array<'a>(
+        &self,
+        schema: &'a JSONSchema,
+        _: &'a Value,
+        instance_value: &'a [Value],
+    ) -> ErrorIterator<'a> {
+        Box::new(
+            instance_value
+                .iter()
+                .zip(self.items.iter())
+                .flat_map(move |(item, validators)| {
+                    validators.iter().flat_map(move |validator| {
+                        helpers::validate(validator.deref(), schema, item)
+                    })
+                })
+                .collect::<Vec<_>>()
+                .into_iter(),
+        )
     }
 }
 
 pub struct ItemsObjectValidator {
     validators: Validators,
 }
-
 impl ItemsObjectValidator {
     #[inline]
     pub(crate) fn compile(schema: &Value, context: &CompilationContext) -> CompilationResult {
         let validators = compile_validators(schema, context)?;
         Ok(Box::new(ItemsObjectValidator { validators }))
     }
 }
-
 impl Validate for ItemsObjectValidator {
-    fn validate<'a>(&self, schema: &'a JSONSchema, instance: &'a Value) -> ErrorIterator<'a> {
-        if let Value::Array(items) = instance {
-            let validate = move |item| {
+    fn name(&self) -> String {
+        format!("items: {}", format_validators(&self.validators))
+    }
+
+    fn is_valid_array(&self, schema: &JSONSchema, _: &Value, instance_value: &[Value]) -> bool {
+        if instance_value.len() > 8 {
+            instance_value.par_iter().all(|item| {
                 self.validators
                     .iter()
-                    .flat_map(|validator| validator.validate(schema, item))
-                    .collect::<Vec<_>>()
-            };
-            let errors: Vec<_> = if items.len() > 8 {
-                items.par_iter().flat_map(validate).collect()
-            } else {
-                self.validators
+                    .all(|validator| helpers::is_valid(validator.deref(), schema, item))
+            })
+        } else {
+            self.validators.iter().all(|validator| {
+                instance_value
                     .iter()
-                    .flat_map(move |validator| {
-                        items
-                            .iter()
-                            .flat_map(move |item| validator.validate(schema, item))
-                    })
-                    .collect()
-            };
-            return Box::new(errors.into_iter());
+                    .all(|item| helpers::is_valid(validator.deref(), schema, item))
```

Comment:
Here validator can be dereferenced before `iter()` so it won't be called multiple times (gave some improvement locally).
There are also other places in the PR where it can be applied

Reply:
Fair point , going to check for other similar occurrences.

Reply:
Addressed all the spotted occurrences in bbe3db6.